ding/reward_model/count_based_model.py

```diff
@@ -0,0 +1,354 @@
+        # otherwise we'll have access to future pixels.
+        if not self._mask_center:
+            hstack += hstack_input
+
+        return vstack, hstack, skip
+
+
+class GatedPixelCNN(nn.Module):
+    """The Gated PixelCNN model."""
+
+    def __init__(
+        # self,
+        # in_channels=1,
+        # out_channels=1,
+        # n_gated=10,
+        # gated_channels=128,
+        # head_channels=32,
+        # sample_fn=None,
+        self,
+        in_channels=1,
+        out_channels=1,
+        n_gated=2,
+        gated_channels=16,
+        head_channels=64,
+        q_level=256,
+    ):
+        """Initializes a new GatedPixelCNN instance.
+        Args:
+            in_channels: The number of input channels.
+            out_channels: The number of output channels.
+            n_gated: The number of gated layers (not including the input layers).
+            gated_channels: The number of channels to use in the gated layers.
+            head_channels: The number of channels to use in the 1x1 convolution blocks
+                in the head after all the gated channels.
+            sample_fn: See the base class.
+        """
+        super().__init__()
+        self._input = GatedPixelCNNLayer(
+            in_channels=in_channels,
+            out_channels=gated_channels,
+            kernel_size=7,
+            mask_center=True,
+        )
+        self._gated_layers = nn.ModuleList(
+            [
+                GatedPixelCNNLayer(
+                    in_channels=gated_channels,
+                    out_channels=gated_channels,
+                    # paper in exploration is 1, default paper is 3
+                    kernel_size=1,
+                    mask_center=False,
+                )
+                for _ in range(n_gated)
+            ]
+        )
+        self._head = nn.Sequential(
+            nn.ReLU(),
+            nn.Conv2d(
+                in_channels=gated_channels, out_channels=head_channels, kernel_size=1
+            ),
+            nn.ReLU(),
+            nn.Conv2d(
+                in_channels=head_channels, out_channels=q_level * out_channels, kernel_size=1
+            ),
+        )
+        self.q_level = q_level
+
+    def forward(self, x):
+        vstack, hstack, skip_connections = self._input(x, x)
+        # print(f'7x7 conv:{skip_connections.shape}')
+        for gated_layer in self._gated_layers:
+            vstack, hstack, skip = gated_layer(vstack, hstack)
+            skip_connections += skip
+            # print(f'3x3 gate:{skip_connections.shape}')
+        # return self._head(skip_connections) # logits
+        logits = self._head(skip_connections)
+        logits, x = logits.permute(0, 2, 3, 1), x.permute(0, 2, 3, 1)
+        # print(f'logits, x:{logits.shape, x.shape}')
+        flattened_logits = torch.reshape(logits, [-1, self.q_level])
+        # print(f'flat logits:{flattened_logits.shape}')
+        target_pixel_loss = torch.reshape(x, [-1])
```

Comment:
why named as target_pixel_loss? I guess x  should be the target pixel.

Reply:
You are right! In the `_probs` function I returned it as `target` as well. I add `_loss` here just to remind me it's used in loss function.